### ISSUE TITLE
CLI says column does not exist on insert sql query

### ISSUE THREAD

**Issue by AppyCat** [cornucopia-rs/cornucopia#74]
Hi I am trying to generate the following query:

--! insert_user()
INSERT INTO users (username)
    VALUES (username)

Error while preparing query "insert_user" [file: "queries/module_3.sql", line: 1] (column "username" does not exist).

I am able to retrieve users but for some reason the above is not working. Also if I put :username that throws another error.

Am using with Postgres.

On another note I was wondering how this may perform over using the Postgres driver natively or vs Diesel or sqlx?

It seems your approach is native as it's essentially taking SQL queries and turning them into functions that engage the native postgres driver directly. Closer to metal vs an ORM, a time saving function generation solution.

**Comment by ianpurton** [comment]
how about adding username as a parameter?

``` 
insert_user(username)
INSERT INTO users (username)
VALUES (username)
```

**Comment by AppyCat** [comment]
That worked perfectly for creating the insert function :)

Got an insert working too as a transaction, but a glitch when trying to accesss it, get the following error:

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: RowCount, cause: None }', src/main.rs:49:48

This line won't work with .vec(), says method not found.

let users = show_users(&transaction).await.unwrap();

Your lib is quite promising, thanks for going in this direction.

Btw, is this approach akin to using the Postgres driver directly, closer to metal?

**Comment by Virgiel** [comment]
The master branch contains the code and documentation for the next version we are working on. You can see the production documentation [here] (https://docs.rs/crate/cornucopia/0.7.4).

> Btw, is this approach akin to using the Postgres driver directly, closer to metal?

 We're not really any closer to metal than `diesel` or `sqlx`, and we're based on a Postgres Rust client `postgres` not the native `libpq` library like diesel. 

> It seems your approach is native as it's essentially taking SQL queries and turning them into functions that engage the native postgres driver directly. Closer to metal vs an ORM, a time saving function generation solution.

I can't speak for @LouisGariepy, but for me it's more of a paradigm difference. To my understanding, diesel generates SQL from Rust, sqlx provides ergonomic helper functions to interact with SQL databases and we generate Rust code from your SQL queries. We try to generate low overhead code that is very ergonomic to use. Essentially, we generate the boilerplate code that nobody wants to write by hand.

**Comment by ianpurton** [comment]
> This line won't work with .vec(), says method not found.
> 
> let users = show_users(&transaction).await.unwrap();
> 

Can you show a code snipper? It should look something like

--! users() { email, name } *
SELECT email, name FROM users;

Notice the '*', this means give me a vec

**Comment by AppyCat** [comment]
Thank you Virgiel, I'll check out that link. I think this approach of generating code from sql is a good value proposition of saving developer time which is gen more cost to businesses than cloud services. 

It also makes arewewebyet more approachable. I like writing sql, so this is a good time saver, one less context switch.

**Comment by AppyCat** [comment]
Thank you for your help. This is the code in main.rs, the .vec throwing the glitch is in the insert transaction. Seems issue was I was going off non master branch code:

`#[rustfmt::skip]
mod cornucopia;
use deadpool_postgres::{Config, Runtime};
use tokio_postgres::NoTls;
use time::OffsetDateTime;

use crate::cornucopia::{
    queries::{
        module_1::show_posts,
        module_2::show_users,
        module_3::insert_user,
    }
};

#[tokio::main]
pub async fn main() {
    let mut cfg = Config::new();
    cfg.user = Some(String::from("postgres"));
    cfg.password = Some(String::from("postgres"));
    cfg.host = Some(String::from("localhost"));
    cfg.port = Some(5432);
    cfg.dbname = Some(String::from("session_test"));
    let pool = cfg.create_pool(Some(Runtime::Tokio1), NoTls).unwrap();
    let mut client = pool.get().await.unwrap();

    {
        let transaction = client.transaction().await.unwrap();
        let created_at_datetime = OffsetDateTime::now_utc();
        insert_user(
            &transaction,
            &"yabbadabbadoo",
            &"fred@flintstones.com",
            &"dino@saursRgreat!",
            &created_at_datetime,
            &created_at_datetime
        )
          .await
          .unwrap();
        let users = show_users(&transaction).vec().await.unwrap();
        println!("{:?}", users);
        transaction.commit().await.unwrap();
    }

    println!("{:?}", show_posts(&client).await.unwrap());
    println!("{:?}", show_users(&client).await.unwrap());
}`

**Comment by AppyCat** [comment]
These are the errors from above, but it seems issue is prob resolved by looking at docs of your latest revision:

error[E0599]: no method named `vec` found for opaque type `impl std::future::Future<Output = Result<(Uuid, String, String, String, OffsetDateTime, OffsetDateTime), tokio_postgres::Error>>` in the current scope
  --> src/main.rs:41:46
   |
41 |         let users = show_users(&transaction).vec().await.unwrap();
   |                                              ^^^ method not found in `impl std::future::Future<Output = Result<(Uuid, String, String, String, OffsetDateTime, OffsetDateTime), tokio_postgres::Error>>`

error[E0599]: no method named `vec` found for opaque type `impl std::future::Future<Output = Result<(i32, String, String, bool), tokio_postgres::Error>>` in the current scope
  --> src/main.rs:46:42
   |
46 |     println!("{:?}", show_posts(&client).vec().await.unwrap());
   |                                          ^^^ method not found in `impl std::future::Future<Output = Result<(i32, String, String, bool), tokio_postgres::Error>>`

error[E0599]: no method named `vec` found for opaque type `impl std::future::Future<Output = Result<(Uuid, String, String, String, OffsetDateTime, OffsetDateTime), tokio_postgres::Error>>` in the current scope
  --> src/main.rs:47:42
   |
47 |     println!("{:?}", show_users(&client).vec().await.unwrap());
   |                                          ^^^ method not found in `impl std::future::Future<Output = Result<(Uuid, String, String, String, OffsetDateTime, OffsetDateTime), tokio_postgres::Error>>`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `postgres-cornucopia` due to 3 previous errors

**Comment by ianpurton** [comment]
```
let users = show_users(&transaction).vec().await.unwrap();
```

I use this

```
let users = show_users(&transaction).await.unwrap();
dbg!(users);
```

I that doesn't work I need to see you .sql file where you specify the query.

**Comment by ianpurton** [comment]
> error[E0599]: no method named `vec` found for opaque type `impl std::future::Future<Output = Result<(Uuid, String, 

You need to await the future and unwrap before calling methods on the result.

i.e. 

```
show_users(&transaction).await.unwrap().vec()
```

But I think you don't need the vec() not in the version I'm using anyway.

**Comment by AppyCat** [comment]
These are my sql files, will try the above now:


--! show_posts()
SELECT
    *
FROM
    posts


--! show_users()
SELECT
    *
FROM
    users


Per your input:


--! insert_user(username, email, password_hash, created_at, updated_at)
INSERT INTO users (username, email, password_hash, created_at, updated_at)
VALUES ($1, $2, $3, $4, $5)

**Comment by ianpurton** [comment]
```
--! show_posts() * <------- STAR
```

You need the star which means I want more than 1 row.

**Comment by AppyCat** [comment]
Aha, works perfectly now :) I like the simplicity & productivity this brings to CRUD, covering 95 to 98% use cases and those last 2 to 5% edge cases one can optimize closer to metal. At the same time cornucopia.rs file provides insight on writing db functions, helpful in the onboarding phase.

Will go over it in more detail tomorrow. Thank you for your help, this is good useful db meta programming

**Comment by AppyCat** [comment]
It seems the performance with your approach may be the postgres client columns in these benchmarks:

https://github.com/diesel-rs/metrics/

**Comment by ianpurton** [comment]
If we look at the [Tech Empower Benchmarks](https://www.techempower.com/benchmarks/#section=test&runid=5871dabb-406e-4bd8-9caf-e8d65ab97888)

The the top performing rust frameworks are generally using tokio-postgres. See ntex https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/Rust/ntex

So I could image cornucopia getting close to this kind of performance.

**Comment by LouisGariepy** [comment]
Thanks @AppyCat and everyone for participating in this issue. If you think your question has been answered I'll go ahead and close this issue :smile: 

Some closing thoughts : 

Sorry for the confusion over versions and docs. While the new upcoming release brings some breaking changes, it makes the query system vastly more powerful while staying as simple, if not simpler for most cases.

Regarding performance, in most benchmarks I have seen, `postgres` and `tokio-postgres` perform admirably, well enough for the vast majority of applications. If we're making comparisons, it seems to be very close to `libpq`, but typically faster than sqlx's driver (as seen in the diesel benchmark you linked). So, cornucopia queries' performance will be similar. To my knowledge, cornucopia doesn't add any significant overhead compared to a good hand-written query.

Since we have the luxury to generate arbitrary code, we avoid many complexities/compromises of other appoaches (complex macros, complex type generics, dyn/boxing). Meanwhile, users get the benefit of a very familiar SQL environment, that can be extended at the user's leasure. As @Virgiel stated, it is a paradigm (SQL-first) more than a way to improve performance (although perf is also good).

Take this with a grain of salt though. At some point I will definitely consider adding proper benchmarks and performance regression testing.

**Comment by AppyCat** [comment]
Thanks Ian, Techempower is partially what led me to revisit Rust per the arewewebyet phrase. Seems the Tokio postgres vs Dieseld generally seem to be a 2 to 1 or 3 to 1 difference at least with a grain of salt on Techempower.

Almost settled on c# until I saw Rust is web ready. I have a lot of Rails exp so recognize the approach similarities with Cornucopia. One final thing not related to this issue, do any of you have a best practice approach to authorization or equivalent to CanCan or Pundit?

I think a Devise & Carrierwave would add value to Rust world, looks like Ian may have it with Barricade and heading in that time saving direction with Rust on Nails. I spent many an hour experimenting with and curating a mix of Rust libs.

The issue is resolved with the feedback, have enough to proceed. Thank you for all the great work on this project Louis and team, I look forward to the updates :)
